integrationtests/agent/adoption_test.go

```diff
@@ -317,71 +102,221 @@ var _ = Describe("Adoption", Label("adopt"), func() {
+				env.assertConfigMapAdopted(g, &cm)
+				g.Expect(cm.Data).To(Equal(map[string]string{"foo": "bar", "key": "value"}))
+				g.Expect(cm.Annotations).To(HaveKeyWithValue("bar", "xzy"))
+				g.Expect(cm.Labels).To(HaveKeyWithValue("foo", "234"))
 			})
 		})
 	})
 
 	When("a bundle adopts a resource that is deployed by another bundle", Label("competing-bundles"), func() {
 		It("should complain about not owning the resource", func() {
-			createBundleDeployment("one", false)
-			waitForConfigMap("cm1")
-			createBundleDeployment("two", true)
-			assertBundleDeployment("one", bundleDeploymentNotOwnedByUs)
-			assertBundleDeployment("two", bundleDeploymentReady)
+			env.createBundleDeployment("one", false)
+			env.waitForConfigMap("cm1")
+			env.createBundleDeployment("two", true)
+			env.assertBundleDeployment("one", env.bundleDeploymentNotOwnedByUs)
+			env.assertBundleDeployment("two", env.bundleDeploymentReady)
 		})
 	})
 })
+
+type adoptEnv struct {
+	namespace string
+	env       *specEnv
+}
+
+func (e adoptEnv) createBundleDeployment(name string, takeOwnership bool) {
+	bundled := v1alpha1.BundleDeployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: clusterNS,
+		},
+		Spec: v1alpha1.BundleDeploymentSpec{
+			DeploymentID: "BundleDeploymentConfigMap",
+			Options: v1alpha1.BundleDeploymentOptions{
+				DefaultNamespace: e.namespace,
+				Helm: &v1alpha1.HelmOptions{
+					TakeOwnership: takeOwnership,
+				},
+			},
+		},
+	}
+
+	err := k8sClient.Create(context.TODO(), &bundled)
+	Expect(err).To(BeNil())
+	Expect(bundled).To(Not(BeNil()))
+	Expect(bundled.Spec.DeploymentID).ToNot(Equal(bundled.Status.AppliedDeploymentID))
+	Expect(bundled.Status.Ready).To(BeFalse())
+	Eventually(func(g Gomega) {
+		g.Expect(k8sClient.Get(
+			context.TODO(),
+			types.NamespacedName{Namespace: clusterNS, Name: name},
+			&bundled,
+		)).To(Succeed())
+		g.Expect(bundled.Status.Ready).To(BeTrue())
+	}).Should(Succeed(), "BundleDeployment not ready: status: %+v", bundled.Status)
+	Expect(bundled.Spec.DeploymentID).To(Equal(bundled.Status.AppliedDeploymentID))
+}
+
+func (e adoptEnv) waitForConfigMap(name string) {
+	time.Sleep(5 * time.Second)
+	Eventually(func() error {
+		_, err := e.env.getConfigMap(name)
+		return err
+	}).Should(Succeed())
+}
+
+func (e adoptEnv) createConfigMap(cm *corev1.ConfigMap) {
+	cm.ObjectMeta.Name = "cm1"
+	cm.ObjectMeta.Namespace = e.namespace
+	Expect(k8sClient.Create(ctx, cm)).To(Succeed())
+	e.waitForConfigMap("cm1")
+}
+
+// assertConfigMap checks that the ConfigMap exists and that it passes the
+// provided validate function.
+func (e adoptEnv) assertConfigMap(validate func(Gomega, corev1.ConfigMap)) {
+	cm := corev1.ConfigMap{}
+	var err error
+	Eventually(func(g Gomega) {
+		err = k8sClient.Get(
+			ctx,
+			types.NamespacedName{Namespace: e.namespace, Name: "cm1"},
+			&cm,
+		)
+		g.Expect(err).ToNot(HaveOccurred())
+		validate(g, cm)
+		g.Expect(err).ToNot(HaveOccurred())
```

Comment:
The expectation on L237 would now be void, since `err` hasn't been updated since L235, unless:
```suggestion
		err = validate(g, cm)
		g.Expect(err).ToNot(HaveOccurred())
```

Reply:
validate runs expectations on it's on using the reference to Gomega, so there's no need to return an exception anymore. I'll remove the line in 237.